### DISCUSSION TITLE
Is there a Prettier equivalent for C#?

### DISCUSSION THREAD

**Discussion by TheMagnificent11** [dotnet/csharplang#7000] [Q&A] [answered]
I'm doing a lot of React development at the moment and I'm loving how [Prettier](https://prettier.io/) is an opinionated way of formatting your code.

Essentially, when you run the **Format Document** command, the file is formatted according to the Prettier rules; it's deterministic so there's essentially only one correct way to write each line of code (I don't think it formats whitespace between lines of code).

It would be great if there was a Roslyn Analyser that could do this for you.

That is, you spec your styling rules in `.editorconfig` and it applies these rules whenever you run the **Format Document** command (in a similar way but more forceful way the way to the **Run Code Cleanup** works in Visual Studio).

Is there anything out there that does this and conforms to the rules of the analyser NuGet packages added to your project?

For example, if I've add the StyleCop NuGet package and create a few overrides to the StyleCop rules in `.editorconfig`, then it reformats the file to conform to the rules.

`dotnet format` partially does this, but doesn't seem to pickup everything.

**Comment by iam3yal** [comment]
Well, both R# and Rider are doing it but they are paid solutions.

**Reply by TheMagnificent11** [reply]
I would prefer a Roslyn analyser solution so that it is editor-agnostic.

That is, you could just run `dotnet format` and it just completely formats all the code in your solution.

**Reply by iam3yal** [reply]
@TheMagnificent11 Both respect `.editorconfig` **BUT** if you're looking for a CLI solution it's completely free, you can check it out [here](https://www.jetbrains.com/help/rider/CleanupCode.html).

**Reply by TheMagnificent11** [reply]
Thanks @eyalalonn I'll look into it.

I want my IDE to behave the same way as these CLI tools such such ’dotnet format‘ and Code Cleanup.

Essentially, I have Roslyn Analysers setup to warn me for everything I want except line length, but very few of these rules can be auto-fixed when you use the format document command.

**Reply by TheMagnificent11** [reply]
So far, I haven't found the Prettier IDE functionality for .Net yet.

Maybe I'll have to use Rider for the trial period to see if it can do it.

However, at the same time, it seems strange to pay when you get this stuff for free when you do JavaScript or TypeScript development.

**Reply by iam3yal** [reply]
@TheMagnificent11 Well, in some ecosystems you get everything for free in others you have to pay and that's the way it is but then there are many available free tools out there for .NET

**Comment by DavidArno** [comment]
[StyleCopAnalyzers](https://github.com/DotNetAnalyzers/StyleCopAnalyzers) is an obvious choice. But I think the project is pretty much dead as there hasn't been a non-beta release for nearly four years and the last beta release was from nearly a year ago. It's also somewhat "old fashioned" in its opinions, so needs a lot of configuring.

As long as you are using at least the .NET 5 SDK, the code style analyzers that were previously part of VS (the IDExxxx rules) are now built into the SDK so can be applied in pipelines etc, so they be the better choice for you. See https://learn.microsoft.com/en-us/visualstudio/code-quality/roslyn-analyzers-overview?view=vs-2022.

**Reply by TheMagnificent11** [reply]
I actually use `StyleCopAnalyzers` and have removed a lot of the "old fashioned" rules.

What I mean is that Prettier is extremely opinionated.

For example, it will format you code to fit the configured max line length if you are over.  If you spread a statement over multiple lines when it can be a single line and still under the max line length, it will format accordingly.

I haven't come across any Roslyn analyser that do anything like that.

**Comment by HaloFour** [comment, marked as answer]
See: https://csharpier.com/

Looks like a small but somewhat active project.

I'm going to go out on a limb that perhaps super-opinionated formatters are less important to C# developers, who probably find the options provider out-of-the-box with Visual Studio or `dotnet format` to be sufficient.  I'm personally not a fan of super-opinionated formatters and find myself fighting with them in projects that do require their use.

**Reply by TheMagnificent11** [reply]
Yes, I would agree.

I have a niche request.

**Reply by TheMagnificent11** [reply]
`CSharpier` looks promising, thanks

**Reply by TheMagnificent11** [reply]
It's almost exactly what I want, but I suppose that's to be expected as it's opinionated.

It conflicts with other style rules that I want to keep.

**Reply by CyrusNajmabadi** [reply]
> It conflicts with other style rules that I want to keep.

> Our ecosystem tends heavily to personal style and not fiat.

:)

**Comment by CyrusNajmabadi** [comment]
Yup. W e get the opposite feedback non-stop. That people want an incredibly configurable formatter that will let them tweak each and every potential decision it makes. Our ecosystem tends heavily to personal style and not fiat.

**Reply by TheMagnificent11** [reply]
Yeah, what I'm asking for is something that does its best to apply fixes to lines of code that fail validation of rules setup in `.editorconfig`.

So, things are still configurable via `.editorconfig` for everyone.

It appears a lot of these issues can be auto-fixed because Visual Studio has "Run Code Cleanup" and the `dotnet format` tool exists.

I guess I'm wondering why more of these can't be auto fixed.

**Reply by CyrusNajmabadi** [reply]
If there are more things you want auto-fixed, please open feature requests for them at dotnet/roslyn.

**Reply by TheMagnificent11** [reply]
How does the auto-fixing work?

Does each Rosyln analyser need to provide a way to fix each rule violation?

I use `StyleCop.Analyzers`, and I'd like to be able to auto-fix a lot of those, but I guess they're not completely under Microsoft's control.  Didn't they originate from the MS Office teams?

Most of the standard `Microsoft.CodeAnalysis.NetAnalyzers` are not as invasive, or maybe I just know them relatively well and they don't come up much.

One that I'd like is `max_line_length`.  I think it's a general `.editorconfig` rules that applies to all files types, but it would be good if that could auto-fix.  However, I guess it then gets into "religious" arguments of things like one parameter per line and where does the `&&` or `||` go when separating if-statements across of multiple lines.

Apparently StyleCop has a line length rule, but it doesn't seem to work when you adjust the length down from the default value.

**Reply by TheMagnificent11** [reply]
Actually, I don't know if it's an analyser concern, but not including the standard `System` using statements that are added when you have `ImplicitUsings` enabled in the C# project.

**Reply by belav** [reply]
> That people want an incredibly configurable formatter that will let them tweak each and every potential decision it makes. Our ecosystem tends heavily to personal style and not fiat.

Is that just because that is what they are used to? When I first tried prettier I hated some of the ways it formatted things. But then I realized the great part about it, there is no discussion, there is no getting used to the new style rules of wherever your next job is, you just get a deterministic result. Run it every time you save and force every PR to be formatted before it can be merged. You no longer have to worry if different IDEs produce different results on a format document/code cleanup. I cringe every time I open a razor file now, it is one of the few types of files left in our huge project at work that doesn't have deterministic formatting. If someone accidentally hits "format document" in one their PR is suddenly a mess of unrelated changes.

I feel like the steady growth of CSharpier shows that there are plenty of people that prefer the opinionated approach. But I can see how c# itself trying to adopt an opinionated approach would probably make a whole lot of people unhappy. I'm guessing CSharpier users are a very small chunk of all c# developers. Full disclosure - I am the author of CSharpier so may be a bit biased.

Probably more of an IDE request, but I'd love a way to run a set of analyzer auto fixes on save. I haven't found anything in my googling to indicate it is possible.

**Reply by KennethHoff** [reply]
I love CSharpier. Can recommend it to everyone who wants consistency in their coding style. Add in StyleCop of y you want even more opinionated formatting :p
